### DISCUSSION TITLE
即將發布的更新內容

### DISCUSSION THREAD

**Discussion by TimLai666** [HazelnutParadise/insyra#6] [Announcements]
這裡會記錄下一個Release會包含的變更內容
版本會以 `上一個版本+` 的形式標註，例如 `v0.0.9+` 為 `v0.0.9` 的下一個版本，但還不確定版號。

**Comment by TimLai666** [comment]
v0.0.9+
- 修正柱狀圖和折線圖未提供 x 軸時會造成圖表錯誤的 bug。
- 修正 `plot` 包可能無法不顯示標籤。
- 新增散佈圖、熱力圖、liquid 圖。
- 提升圖表輸出畫質。
- 圖例改為預設顯示在圖表下方，讓整體呈現更協調。
- 新增 `gplot` 包，可畫直條圖、直方圖。
- 新增 `lpgen` 包，可生成線性規劃 `.lp` 檔。
- 新增 `lp` 包，用於線性規劃。
- 修正 DataTable 在某些情況下不會重置索引的問題。
- 將單元測試根據功能拆成各個部分。 by @ksw2000 
- DataList 及 DataTable 的時間戳改為併發安全。 by @mudream4869

**Reply by TimLai666** [reply]
Released

**Comment by TimLai666** [comment]
v0.0.10+
- `stats`: 不再需要對結果斷言。
- `plot`: 
  - 修正圖表渲染失敗時沒有如期使用線上服務渲染的 bug。
  - 新增 wordcloud、雷達圖、箱形圖、gauge 圖。
- 新增 `csvxl` 包，可將 CSV 轉為 Excel 工作表，或將 Excel 工作表轉為 CSV。
- DataList: 修正 ReplaceAll、ReplaceFirst、ReplaceLast 方法的 bug。
- 加入更多單元測試。

**Reply by TimLai666** [reply]
Released

**Comment by TimLai666** [comment]
v0.0.11+
- 新增 `py` 包，可執行 Python 程式碼，並預裝多個常用套件！（目前僅支援 MacOS、Linux）
- DataList:
  - 新增 `IsEqualTo` 方法，用於比較 DataList 中的資料是否與另一個 DataList 相同。
  - 新增 `IsTheSameAs` 方法，用於比較 DataList 是否與另一個 DataList 完全相同。
  - 新增 `Counter` 方法，可計算每個元素出現次數並將結果回傳為 map。
  - 修正舊時間戳可能會覆蓋新時間戳的問題。
  - 修正 `ClearStrings` 遇到 cpu 核心數大於資料長度會引發錯誤的問題。By @ksw2000 
- DataTable:
  - 修正舊時間戳可能會覆蓋新時間戳的問題。
  - 修正當 Column 尚不存在時，`AppendRowsByIndex` 不會按照字母順序插入的問題。
  - 將 `AppendRowsByIndex` 改名為 `AppendRowsByColumnIndex`；`AppendRowsByName` 改為 `AppendRowsByColumnName`，避免混淆。
- `csvxl`: 支援 utf-8 和 big5，預設使用 utf-8。
- `plot`: 
  - 新增漏斗圖、ThemeRiver、桑基圖。
  - 輸入資料的屬性名稱統一為 `Data`。
- `gplot`:
  - 輸入資料的屬性名稱統一為 `Data`。
  - 新增一元函數圖。
- `lp`:
  - 將 `SolveLPWithGLPK` 改名為 `SolveFromFile`。
  - 新增 `SolveModel` 函數，可直接解 `lpgen` 的模型。
  - 優化 GLPK 的初始化安裝，在 macOS 和 Linux 上不再依賴 brew 和 sudo。
- 新增更多單元測試。Thanks @ksw2000 

**此版本測試覆蓋率：34.5%**

**Reply by TimLai666** [reply]
Released

**Comment by TimLai666** [comment]
v0.0.12+
- `py`:
  - 支援 Windows。
  - 優化 `RunCodef` 內部邏輯，避免 Python 變數名稱衝突的可能。
- `lpgen`:
  - 修正 `.lp` 檔的格式問題。
  - 建立模型時支援鏈式呼叫方法。
- 增加單元測試。

**此版本測試覆蓋率：35.6%**

**Reply by TimLai666** [reply]
Released

**Comment by TimLai666** [comment]
v0.0.13+
- DataTable:
  - 將函數及方法名稱中的 Column 改為 Col。
  - ColIndex 支援小寫字母。
  - 新增 `Counter` 方法。
- 加入 GolangCI 及 Govulncheck，優化程式碼。 Thanks @adrian-lin-1-0-0 @vn7n24fzkq 
- 加入更多單元測試。

**此版本測試覆蓋率：38.7%**

**Reply by TimLai666** [reply]
Released

**Comment by TimLai666** [comment]
v0.0.14+
- `lpgen`:
  - 新增能把 LINGO 模式轉為標準 lp 格式的功能。
- `stats`:
  - 新增線性回歸（`LinearRegression`）。
  - 新增單元測試。

**此版本主套件包測試覆蓋率：38.7%**

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.0.15+
- 更新 Go 版本至 1.23。
- 更新依賴套件，解決安全問題。
- 新增 `datafetch` 包，可直接獲取 Google Maps 商家評論資料。
- 新增 `isr` 包，提供語法糖。
- `csvxl`:
  - 新增 `EachExcelToCsv` 和 `EachCsvToOneExcel` 來批量轉換資料夾內的檔案。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.1.4+
- DataTable:
  - 新增轉換成 JSON 的功能，包含 `ToJSON` 和 `ToJSON_Bytes` 方法。
  - 修正使用 `AppendCols` 之後，DataTable 仍與外部 DataList 保持連動的錯誤機制。
- DataList:
  - 調整 `Mode` 方法，以符合眾數定義。
- `isr`:
  - 提供更多語法糖。
- `datafetch`:
  - 修正一些 bugs。
- `py`:
  - 調整環境初始化邏輯。
  - 增加 `RunFile` 及 `RunFilef` 以支援執行 .py 檔。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.1.9+
- DataTable:
  - `Show`, `ShowTypes`: 美化顯示。
  - 新增 `ShowRange`, `ShowTypesRange`，可指定顯示的列數。
  - 新增 `Summary`，可顯示敘述統計結果。
  - 新增 `Map`，可將所有元素應用自訂計算並回傳新的DataTable。
- DataList:
  - 新增 `Show`, `ShowTypes` 方法來顯示內容和資料型態。
  - 新增 `ShowRange`，可指定顯示的列數。
  - 新增 `Summary`，可顯示敘述統計結果。
  - 新增 `Map`，可將所有元素應用自訂計算並回傳新的DataList。
- `isr`:
  - 更改 `DT` 和 `DL` 的初始化方式，不再需要使用 `{}` 初始化結構體。
  - 修正CSV載入設定的欄位名稱為 `InputOps`。
- 新增可搭配 Gorm 操作 SQL 資料庫的功能。
- 全面改寫 `stats` 包。
- 重寫 DataTable 說明文件。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.1+
- 修正資料中若有 `nil` 會無法創建 DataList 的問題。
- `stats`:
  - 加入迴歸、相關矩陣等。
  - 新增更多測試，並修正多個算法。
- DataTable:
  - 新增 `SetColNameByIndex`、`SetColNameByNumber`、`ChangeColName`、`GetColNameByNumber`、`GetColByName`。
  - 新增 `ChangeRowName`、`GetRowByName`。
  - 新增 `SwapColsByName`、`SwapColsByIndex`、`SwapColsByNumber`，來交換 DataTable 中兩欄的位置。
  - 新增 `SwapRowsByIndex`、`SwapRowsByName`，來交換 DataTable 中兩列的位置。
- 加入 Column Calculation Language，可用於在 DataTable 中新增計算欄。
- 加入錯誤蒐集與處理機制，可用於統計軟體開發。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.3+
- 啟用版本名：從這個版本開始，每個中型版本（v0.x）都會有自己的版本名，以台灣景點命名。v0.2.x的版本名為 **駁二（Pier-2）**。
- `stats`:
  - 迴歸分析的函數統一改為依變數在前、自變數在後的型式。
- `isr`:
  - 現在可使用 `DT.From(nil)` 創建空 DataTable。
  - 新增 `Rows` 和 `Cols` 兩個類型別名。
  - `JSON` 結構體新增 `Bytes` 欄位。
  - `CSV` 結構體新增 `String` 欄位。
- DataList:
  - 全面棄用互斥鎖，改用 `AtomicDo` 來確保原子性。
  - 刪除記憶體重組機制。
- DataTable:
  - 新增 `LoadFromJSON_Bytes`、`ColNamesToFirstRow`、`DropColNames`、`ColNames`、`RowNamesToFirstCol`、`DropRowNames`、`RowNames`。
  - 新增 `GetColNameByIndex`，更方便取得欄名。
  - 新增 `GetColIndexByName`、`GetColIndexByNumber`、`GetColNumberByName`，加強index、number、name之間的互動。
- 將 DataTable 的 `LoadFromJSON`、`LoadFromJSON_Bytes` 和 `LoadFromCSV` 方法改為包級函數 `ReadJSON`、`ReadJSON_Bytes` 和 `ReadCSV`。
- 新增 `ReadCSV_String`。
- 重新編寫 DataList 文檔。
- 新增 `mkt` 包，提供 RFM 等行銷相關數據分析。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.4+
- DataTable:
  - 新增 `Clone` 方法，可創建資料副本。
  - GetCol 系列方法在錯誤時統一回傳 nil。
  - 新增 `SortBy` 方法，用以排序 DataTable，支援多層排序。
  - 更新 JSON 序列化庫為 goccy/go-json，提升性能。
  - 新增 `SimpleRandomSample` 方法，用於隨機抽樣。
  - 新增 `Headers` 方法作為 `ColNames` 的別名。
  - 新增 `SetColNames` 和 `SetHeaders` 來設置一整列欄名。
  - 新增 `SetRowNames` 來設置一整行列名。
  - 完善並發安全性。
- DataList:
  - 確保 `Sort` 的排序機制。
  - 完善並發安全性。
- `mkt`:
  - 根據時間尺度來計算 RFM。
  - `RFM` 函數在日期解析錯誤時不再停止計算，而是跳過該列。
  - 新增 `CustomerActivityIndex` 函數，用於計算顧客活躍性指標，了解顧客活躍度變化趨勢。也可使用 `CAI` 別名呼叫。
- `isr`:
  - 提供新函數 `UseDL` 和 `UseDT` 來取代名稱不夠直覺的 `PtrDL` 和 `PtrDt`。
  - 完善並發安全性。
  - 新增單元測試。
- `datafetch`:
  - 更新 JSON 序列化庫為 goccy/go-json，提升性能。
- `plot`:
  - 完善並發安全性。
- `py`:
  - 更新 JSON 序列化庫為 goccy/go-json，提升性能。
- `stats`:
  - 完善並發安全性。
- `parallel`:
  - `Run` 方法會根據傳入的函數是否有回傳值來決定要不要處理結果，以提升性能。
  - 新增 `AwaitNoResult` 方法，用於不需回傳結果的任務群組。
- utils:
  - 不再對外暴露 `TruncateString`、`FormatValue`、`ColorText`。
  - 新增 `SortTimes`，用於排序時間。
- 新增 `Dangerously_TurnOffThreadSafety` 方法，可關閉線程安全。
- 將 Go 版本更新至 1.24，並更新所有依賴。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.5+
- DataList:
  - 初始化或Append時只輾平切片，不再輾平陣列。
- DataTable:
  - 新增 `DropColsContain` 與 `DropRowsContain`，用於去除包含特定值的欄或列。
  - 新增 `DropColsContainExcelNA`、`DropRowsContainExcelNA`，用於快速清除 Excel 的錯誤值。
- `csvxl`:
  - 新增 `DetectEncoding` ，自動偵測編碼格式。
  - 行為修正：Excel工作表轉換為CSV時，只保留工作表中可見的紀錄（未被篩選排除）。
- `stats`:
  - 卡方檢定結果提供 `Show` 方法，可輕鬆顯示檢定結果。
  - 將 `ChiSquareGoodnessOfFit` 的輸入改為類別變數而非頻次。
- `gplot`:
  - 新增 Heatmap、Step Chart、Error Bars。
- `mkt`:
  - 將 CAI 的計算修正為百分比。
  - 將 RFM 結果的 RFM_Score 欄改為 R_Score、F_Score、M_Score 的總和。
- `py`:
  - 改用 UV 管理環境。
- DataTable 和 DataList 的 `Show` 會在最後加上換行。
- 新增包級 `Show` 函數，用於呼叫物件的 `ShowRange` 方法並自動處理 nil。
- 性能提升。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.6+
- DataTable:
  - 新增 `To2DSlice`，可將 DataTable 轉成二維切片。
- `py`:
  - 更精確地固定 Python 版本。
  - 預裝 Polars。
  - Python 中呼叫 `insyra_return` 之後自動結束 Python 程式，避免多次呼叫導致結果被篡改。
  - Run 系列函數不再固定回傳 map，而是由使用者傳入結構體指標來綁定執行結果。
  - Run 系列函數會回傳 Python 程式錯誤，讓錯誤處理更優雅。
  - Go 變數傳給 Python 時，自動處理布林值和 nil、DataList 自動轉換為 pd.Series、DataTable 自動轉成 pd.DataFrame。
  - 可在 Run 系列函數中使用的 Python 工具函數或屬性:
    - `insyra_return` 改為 `insyra.Return`，舊寫法仍保留。
    - `insyra.execution_id`: 在 Python 中取得該次 Python 程式執行的執行 ID。
- 新增 `Slice2DToDataTable`，來將二維切片轉為 DataTable。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.7+
- `mkt`:
  - 修正 `RFM` 日期格式分鐘的處理。
  - 修正 `CAI` 計算邏輯，僅考慮至少四次交易的顧客。
- `isr`:
  - 支援使用二維陣列創建 DT。
- DataTable:
 - 新增 `ToJSON_String` 和 `ToMap`。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.8+
- 新增關閉彩色輸出的功能。
- 在 `README.md` 中說明設定用法。

### Read 系列函數重大調整
- `ReadCSV` → `ReadCSV_File`
- `ReadJSON` → `ReadJSON_File`
- `ReadJSON_Bytes` → `ReadJSON`
- 新的 `ReadJSON` 函數可接受任何可解析為 JSON 的資料型別，自動解析 JSON。
- 新增 `ReadSlice2D` 函數，為 `Slice2DToDataTable` 的別名。

### CCL 語法更新
- 提供新的 `[colIndex]` 和 `['colName']` 語法用於指定欄，解決欄索引可能和函式撞名的問題。
- 提升 CCL 執行效能。
- 不再僅限於表達式，可搭配 DataTable 新增的 `ExecuteCCL` 方法執行欄賦值指令。
- 可使用 `NEW('colName')=...` 創建新的欄。
- 多行指令使用分號或換行分隔。
- 改進鏈式比較，解決 `>=`、`<=` 可能出錯的問題。

## DataTable
- 新增 `FilterRows`，用於根據條件篩選列。
- 新增 `FilterCols`，用於根據條件篩選欄。
- 修改 filter 系列函數名稱，使其更容易理解。
- update 系列函數回傳 DataTable 指標，支援鏈式呼叫。
- 新增 `ExecuteCCL`，用於在 DataTable 上執行多行 CCL 指令。
- 新增 `EditColByIndexUsingCCL`、`EditColByNameUsingCCL`，用於編輯現有欄。

## DataList
- `Update`、`InsertAt` 函數回傳 DataList 指標，支援鏈式呼叫。

## csvxl
- 改進 `ExcelToCsv` 的演算法，並修正 CSV 檔數量計算。

## stats
- 移除尚不穩定的因素分析程式碼。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.9+
- `ReadCSV_File` 現在可以指定編碼，若無指定則自動偵測，以支援 utf-8 以外的格式。
- 新增 `ReadExcelSheet`，可直接將 Excel 工作表讀取成 DataTable。
- 更新依賴套件。

## DataList
- 新增 `ReplaceNaNsAndNilsWith`、`ReplaceNilsWith`、`ReplaceNaNsWith`、`ClearNilsAndNaNs`、`ClearNils`。

## DataTable
- 新增 `GetRowIndexByName`、`DropColsContainNaN`、`DropRowsContainNaN`。
- 調整 Drop 系列方法的函數名稱和回傳值，使命名更統一並方便鏈式呼叫。
- `GetRowIndexByName` 和 `GetRowNameByIndex` 方法現在會回傳布林值，更方便使用。

## `isr`
- 新增 `Of` 方法作為 `From` 的別名，提供更短的語法糖。
- 新增 `DT.CCL` 方法以執行 CCL 語句。
- `CSV_inOpts` 提供編碼選項。
- 新增 `Excel` 結構體，與 `DT.Of` 或 `DT.From` 搭配可將 Excel 工作表匯入 DataTable。

## `mkt`
- 修正 `RFM` 的 F 計算邏輯，同會員同一時間尺度下只計算一次。
- 修正 `RFM` 和 `CAI` 的時間差計算，使其符合給定的時間尺度。

## `csvxl`
- 所有函數改為回傳錯誤，讓錯誤處理更容易。
- 將 `DetectEncoding` 移到 insyra 核心包。
- 新增 `ReadCsvToString`，可將 CSV 檔讀取成字串。

## `plot`
- 修正 `SavePNG` 的內部邏輯。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.10+

## `parquet`

此次更新新增的包，用於讀取 .parquet 檔，並且能將 DataTable 轉為 parquet。

## DataList
- `ReplaceFirst`、`ReplaceLast`、`ReplaceAll` 現在都會回傳修改後的 DataList，方便鏈式呼叫。
- Replace 系列方法支援 NaN。
- Drop 系列方法支援 NaN。
- `DropIfContains` 修正為接收一個子字串。
- Find 系列方法支援 NaN。
- `Count` 支援 NaN。
- 新增 `Concat` 方法，可和另一個 DataLsit 連接並回傳新 DataList。
- 新增 `AppendDataList`，用於直接將另一個 DataList 加進目前 DataList。
- `Show` 方法字串現在會顯示單引號 `'`，數字會靠右對齊。

## DataTable
- `DropColsContain`、`DropRowsContain` 支援 NaN。
- 新增 `Replace`、`ReplaceNaNsWith`、`ReplaceNilsWith`、`ReplaceNaNsAndNilsWith`、`ReplaceInRow`、`ReplaceNaNsInRow`、`ReplaceNilsInRow`、`ReplaceNaNsAndNilsInRow`、`ReplaceInCol`、`ReplaceNaNsInCol`、`ReplaceNilsInCol`、`ReplaceNaNsAndNilsInCol`。
- `Count` 支援 NaN。
- `SetRowNameByIndex` 現在可鏈式呼叫。
- 新增 `Merge` 方法，用於合併兩個 DataTable，並回傳合併好的新 DataTable。
- 修正 `Show` 方法，讓欄名很長的極端情況也能正常顯示。並且字串現在會顯示單引號 `'`，數字會靠右對齊。
- 新增 `NumRows` 和 `NumCols`。

## CCL
- 支援 nil 型別。
- 新增 `ISNA` 和 `IFNA` 函數，用於檢查 NaN 和 Excel 錯誤。
- 新增 `.` 運算子來指定欄中的某一列。
- 新增 `@` 運算子來代表所有欄，`@.0` 可指定第一列。
- 新增聚合函數 `SUM`、`AVG`、`COUNT`、`MAX`、`MIN`。
- 新增列索引運算子 `#`，用於取得目前列索引。
- 新增範圍運算子 `:`。

## `plot`
- 由於 go-echarts 的設計缺陷，我們決定移除液體圖。
- 重新設計整個包，使其更容易使用。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.11+

## `py`
- 將接收 Python 結果的伺服器修正為只聽 localhost。
- Python 端回傳 DataFrame 或 Series 時，Go 端現在可由 DataTable 或 DataList 接收。一維串列或元組也可以用 DataList 接收；二維串列或字典或字典串列可以用 DataTable 接收。
- 新增 `PipList` 和 `PipFreeze`，用來取得目前 Python 環境中的套件清單。
- `PipInstall` 和 `PipUninstall` 現在會回傳錯誤。
- 新增 `RunCodeContext`、`RunCodefContext`、`RunFileContext`、`RunFilefContext` 方便使用者取消執行 Python 程式碼。
- 新增 `RunCodeWithTimeout` 自動在超時後取消執行 Python。
- 環境初始化失敗將回傳錯誤，不再 Fatal（伺服器初始化失敗除外）。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.12+

- 將所有指令執行改成不彈出終端機視窗。
- 改善文件，提升易讀性。

## 錯誤處理
1. 新增 ErrorInfo 結構 (error_buffer.go)
```go
  type ErrorInfo struct {
      Level       LogLevel
      PackageName string
      FuncName    string
      Message     string
      Timestamp   time.Time
  }
```
  - 實作 error 介面的 Error() 方法
  - 新增 LogLevel.String() 方法

  2. 全域錯誤緩衝區增強 (error_buffer.go)

  新增便捷方法：
  - HasError() / HasErrorAboveLevel(level) - 檢查是否有錯誤
  - PeekError(mode) - 查看錯誤但不移除
  - GetAllErrors() - 取得所有錯誤
  - GetErrorsByLevel(level) / GetErrorsByPackage(pkg) - 篩選錯誤
  - PopAllErrors() / PopErrorInfo(mode) - 取出並移除錯誤

  3. 實例級別錯誤追蹤 (datatable.go, datalist.go)

  - 新增 lastError *ErrorInfo 欄位
  - 新增 Err() - 取得最後一個錯誤
  - 新增 ClearErr() - 清除錯誤並返回自身（支持鏈式呼叫）
  - 新增 setError() - 設定實例錯誤
  - 新增 warn() 輔助方法 - 同時呼叫 LogWarning 和 setError

## CCL
- 修正像 IF(#>0, ['date'].(#-1) - ['date'], NULL) 這類表達式時，IF 的參數會被 eager evaluate 的問題。
- 日期及日期字串可以計算（回傳 time.Duration）。
- 新增 `DAY`、`HOUR`、`MINUTE`、`SECOND` 函數。

## DataList
- `Append` 方法回傳 DataList，方便鏈式呼叫。

## DataTable
- 使用新的 `BiIndex` 結構來儲存列名，提升查詢速度。
- `Merge` 支援 left join 和 right join，並且 on 參數可接受兩個欄名（左表與右表）。
- 提升欄索引和數字索引的轉換性能。

## `plot`
- 提供精美的預設配色。

## `py`
- 改為使用暫存檔執行 Python 程式碼，避免命令列指令長度限制。
- 回傳 Python 結果改用 IPC（Windows Named Pipe / Unix Domain Socket）。

## `datafetch`
- 新增 Yahoo Finance 資料，使用 [github.com/wnjoon/go-yfinance](https://github.com/wnjoon/go-yfinance)。

**Reply by admin** [reply]
released

**Comment by admin** [comment]
v0.2.13+
- 新增 `pd` 包以實現 Insyra 和 [gpandas](https://github.com/apoplexi24/gpandas) 之間的互通性。
- 新增 `engine` 將實用的內部演算法和結構導出，供其它開發者靈活應用。

## `py`
- `PipInstall`、`PipUninstall`、`PipList`、`PipFreeze` 執行時現在都會隱藏視窗。

## `gplot`
- 更改 API，將資料改為函數的參數。

**Reply by admin** [reply]
released

**Comment by admin** [comment]
v0.2.14+
- 新增 Insyra 的 Agent Skills（位於 skills 資料夾下）。
- 新增 Insyra CLI。
- 新增主題引導式文件。

## DataList
  - `Rank` 函數接受可選的 `ascending` 參數。

**Reply by TimLai666** [reply]
released

**Comment by TimLai666** [comment]
v0.2.15+
- 修正cli主程式在上一版被 gitignore 掉的問題。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.16+

## Mutex + fast-goid AtomicActor rewrite
以 sync.Mutex + petermattis/goid 設計取代了 channel-actor + runtime.Stack 的 getGID 方法。AtomicDo 操作速度提升約 100 倍，統計方法速度提升 5-24 倍。

## 新套件
新增 `finance` 包，用於高精度金融計算。

## `stats`
- 內部重構，盡量使用 Gonum 避免重複造輪子，並加強對齊 R。
- 現在會明確回傳錯誤，而不是 LogWarning。
- 新增集群方法。
- 新增因素分析。

## CCL
- 擴充多個數學函數，讓 CCL 像 Excel 公式一樣方便。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.17+

- 為 DataList & DataTable 新增時間序列/視窗變換 (Shift, Diff, PctChange, CumSum/Prod/Max/Min, unified Rolling & Expanding)。

## DataTable
- 把SQL 讀寫升級成「批次 INSERT、schema 只查一次、支援 context 取消、ReadSQLStream 串流大表」，並補齊參數化查詢、Schema、IndexCol、ParseDates、DType 等 pandas 級選項。
-  新增 GroupBy / Aggregate / AggregateAll / Count，15 種 AggregateOp（Sum/Mean/Median/Min/Max/Count/CountAll/Stdev/StdevP/Var/VarP/First/Last/NUnique/Custom）。
- 加了 Pivot / Unpivot 給 DataTable 做長↔寬資料轉換（含 isr 包裝、CLI 指令、文件），並把名稱誤導的 ConvertLongDataToWide 標 deprecated。

## `isr`
- 新增 `(*dt).GroupBy()` 方法。

## CLI / REPL / DSL
- 新增 `groupby` 指令。
- `load` 及 `save` 指令可加參數控制是否將第一列當欄名以及是否第一欄當列名，`save` 指令可加參數控制是否加 bom。
- DSL session 支援自訂環境儲存位置（新增 env.Manager 型別），cli/env 重構為 Manager-based，環境位置不再寫死於 ~/.insyra。
### 資料庫操作指令 
- 新指令 `db connect <name> <dsn>` — 開啟具名資料庫連線，存活於 session。
- 新指令 `db list` — 列出目前所有連線（dialect + masked DSN，密碼顯示 `***`）。
- 新指令 `db tables <name> [schema <s>]` — 列出該連線下的所有 table。
- 新指令 `db disconnect <name>` — 關閉指定連線
- load 加 sql 子形式 — `load sql <conn> <table> [where/order/limit/offset/cols/schema/indexcol/parsedates] [as $var]`。
- `load sql <conn> query "<SQL>"` — 自訂查詢，支援 params <v1> <v2> ... 綁參數。
- save 加 sql 子形式 — `save <var> sql <conn> <table> [if-exists fail|replace|append] [batch N] [schema <s>] [rownames]`。
- REPL 結束自動關閉資料庫連線 — `commands.CloseAllDBConns(ctx)` 在 REPL 退出時 defer 執行。
- 三套純 Go driver 編進 binary — SQLite（modernc 經 glebarez）、MySQL（go-sql-driver）、Postgres（pgx），CGO_ENABLED=0 也能 build。
- DSN 格式 — `sqlite:<path-or-:memory:>`、`mysql://... 或 mysql:user:pass@tcp(...)/db`、`postgres://... 或 postgres:host=... dbname=...`。

## `stats`
- 新增無母數檢定（Wilcoxon, Mann-Whitney U, Kruskal-Wallis, Friedman）。

## `mkt`
- 新增購物籃分析。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.18+

這是 **Pier-2(駁二)** 的最後一個版本，下個版本將進入 **v0.3 Huashan(華山)**。

## DataList / DataTable
- 新增更多缺失值插補方法。
- 補上抽樣、比例抽樣、洗牌與 train/test split。
- 為機器學習預處理新增分類編碼工具(one-hot, label, ordinal)。
- 新增可程式化的資料摘要 Describe() API 與 CLI 指令，支援數值、類別、分組統計與自訂分位數。
- 新增可重複使用的特徵縮放器（StandardScaler / MinMaxScaler / RobustScaler / MaxAbsScaler），支援在訓練集 fit、用同一組參數轉換測試集的無洩漏（no-leakage）ML 預處理流程，並提供對應的 stateful CLI scale 指令。

## `stats`
- 加入廣義線性模型 (Logistic / Poisson / generic GLM)。

**Reply by TimLai666** [reply]
released

**Comment by admin** [comment]
v0.2.19+

* 推出新套件包 `quant`。
* 新增 `insyra.AtomicDoAll`，可一次原子鎖定多個 DataList/DataTable。
* `DataList`/`DataTable` 新增 `ShowTo`/`ShowRangeTo`/`SummaryTo`，可輸出到任意 `io.Writer`。
* CSV/JSON 讀取改為欄位級型別推斷，整數欄位讀為 `int64`（破壞性）。

## Core

* 修正 actor 模型跨-instance 資料競態（trust-zone）。
* `AppendCols`/`UpdateCol`/`UpdateColByNumber` 改為插入時複製，不再別名外部 `DataList`。
* 全庫稽核，修正 80+ 個確認缺陷。

## CLI

* `show`/`describe`/`summary` 輸出改走 `ctx.Output`。 

## `stats`

* NormCDF/NormPPF 轉公開。
* 分位數統一為 R type-7（破壞性）。

## `mkt`

* `RFM` 的 Recency 改以資料集最新交易日為基準。

## `finance`

* 重新定位為財務正確、Excel-familiar API。
* `TBillEq` 對到期逾 182 天改用 Treasury/SIA 公式（破壞性）。

## `datafetch`

* 新增 TWGeocoding 台灣逆地理編碼 fetcher,把經緯度座標轉成縣市/鄉鎮/村里,支援單點與批次查詢、rate-limit 處理與可選快取。

**Reply by admin** [reply]
released

**Comment by admin** [comment]
下一版的更新內容改記在 repo 裡了。

之後每一版要發布的變更，會直接寫進 repo 根目錄的 CHANGELOG.md（英文）與 CHANGELOG_TW.md（繁體中文），跟造成該變更的 PR 一起進來，不再累積在這則討論。

已發布的版本仍然看 Releases：https://github.com/HazelnutParadise/insyra/releases

這則討論保留作為歷史紀錄。

---

Pending release notes have moved into the repository. Upcoming changes are now recorded in CHANGELOG.md and CHANGELOG_TW.md at the repo root, in the same pull request that causes them, rather than accumulating in this discussion. Published releases stay on the Releases page. This discussion remains as a record.
